src/compute-types/src/plan/mod.rs

```diff
@@ -1846,37 +1845,35 @@ This is not expected to cause incorrect results, but could indicate a performanc
             Self::refine_union_negate_consolidation(&mut dataflow);
         }
 
-        if enable_monotonic_oneshot_selects {
```

Comment:
Why don't we continue guarding this by `desc.is_single_time()?` We can `assert!` this in `refine_single_time_operator_selection` and `refine_single_time_consolidation` and we don't need to do anything in the `Optimize` implementations.

Reply:
We already test for `desc.is_single_time()` in both `refine_single_time_operator_selection` and `refine_single_time_consolidation`:
https://github.com/MaterializeInc/materialize/blob/c33083cce4e7da390aa854737b8e242afedede72/src/compute-types/src/plan/mod.rs#L2064-L2068
https://github.com/MaterializeInc/materialize/blob/c33083cce4e7da390aa854737b8e242afedede72/src/compute-types/src/plan/mod.rs#L2117-L2124
The reason that these are not `assert!` is that this planning code is shared for both single-time dataflows and "vanilla" dataflows. So the checks are there to ensure that these refinements are only applied when we see dataflow that is actually single-time.

What I am not sure is whether you intended that the calls in the `Optimize` implementations be restricted only to "vanilla" dataflows? If so, then the code creating the `DataflowDescription` needs to ensure that the `as_of` and the `until` are set correctly to avoid the single-time dataflow case. The new `assert!` statements are documenting this invariant, but happy to drop them if the code paths would be also intended for single-time dataflows.

Reply:
> The reason that these are not `assert!` is that this planning code is shared for both single-time dataflows and "vanilla" dataflows

I might be getting something wrong, but the only caller of these two methods is `finalize_dataflow`.

What we do now:

1. Pass hard-coded `enable_monotonic_oneshot_selects: false` in single-dataflow `Optimize~` structs.
2. Condition one-shot specialization in `finalize_dataflow` call based on `enable_monotonic_oneshot_selects`.
3. Have a short-circuit `return` if we are not `is_single_time` in `refine_single_time_~` methods.

What you are proposing to do:
1. Pass hard-coded `enable_monotonic_oneshot_selects: false` in single-dataflow `Optimize~` structs.
2. Run one-shot specialization unconditionally in `finalize_dataflow`.
3. Have a short-circuit `return` if we are not `is_single_time` in `refine_single_time_~` methods.

What I'm suggesting to do:
1. Don't do anything `Optimize~` structs, just remove the feature flag plumbing.
2. Run the code conditionally in `finalize_dataflow` if only if `is_single_time()` is true.
3. Assert that `is_single_time()` is true in `refine_single_time_~` method calls.

Reply:
I'm also fine with just removing from the `Optimize~` structs, as the `finalize_dataflow` changes that you are proposing will not do any work if we are not in a single time dataflow.

The asserting that we are not in a single time dataflow in `Optimize~` is feels somewhat weird because this is guaranteed by the `Optimize~` implementation itself.

Reply:
Yeah, sorry, the feature flag value (bound to the `OptimizerConfig` field) was passed through only in `OptimizePeek` which is still on my WIP branch. I can see how without looking at all all `Optimize~` structs this can be confusing.